### PR TITLE
objinfo: tell when the object file comes from a different OCaml version

### PR DESCRIPTION
**Pull request by gasche**
When passed an object file that comes from a different OCaml version,
objinfo will currently just detect that the magic number is not what
it expects, and claim it is not a valid file. This is irritating.

The current patch implements some more advanced handling of magic
number in a new Misc.Magic_number module, inspired by js_of_ocaml
Misc.MagicNumber module
  https://github.com/ocsigen/js_of_ocaml/blob/151b811/compiler/util.cppo.ml#L277-L347
that lets us easily tell the difference between "not a valid magic number"
and "a valid magic number, but with another version".

The error message output by objinfo now looks like this:

```
$ ./objinfo ./ocamldep.cmo
File ./ocamldep.cmo
Wrong magic number:
this tool only supports object files produced by compiler version
    4.03.0+dev12-2015-11-20,
and this object file (cmo) seems to come from an older version.
```

### CODE REVIEW

**Comment by gasche** [comment]
Once we have some logic in the compiler codebase that understand magic numbers, we can use it in other places than (ocaml)objinfo. In particular, when the compiler itself reads a compiled object file, it currently has the same irritating behaviour of not recognizing that object file from other OCaml versions are a bit more valid than random text files with the wrong extension, and we could easily fix that. Depending on the feedback on this PR, I may try to do this it as well.

There is no Changes entry for this one. I'm not quite sure where I should put it, and if I understand correctly `objinfo` is not currently installed along with the compiler so it's arguably not an user-facing change that needs to be documented.

**Comment by damiendoligez** [comment]
Look for `read_cmi` in `typing/cmi_format.ml`.

**Comment by gasche** [comment]
Oh, I hadn't seen that the `cmi` reading part had this same logic. But I think this could be generalized to other areas of the compiler (all those that read object files), for example `driver/pparse.ml` (I'm not familiar with the workflow of the people passing binary AST files, but I would assume that version mismatches do happen), `asmcomp/compilenv`, `bytecomp/byte{link,librarian}`, `dynlink`, and `toplevel/topdirs` (load directive).

**Comment by damiendoligez** [comment]
I agree it's a good idea to generalize it, especially if you share the implementation.

**Comment by whitequark** [comment]
It is installed:

```
$ which ocamlobjinfo
/home/whitequark/.opam/4.02.3/bin/ocamlobjinfo
```

**Comment by gasche** [comment]
So I started working a bit more on this PR, but:
- I'd rather wait for the Changelog to be branched off before adding a Changes entry (as this is not going to be in 4.03, I suppose)
- there is some non-trivial interaction between extending this work to `pparse.ml` and #351 , so I would rather do the extension after #351 has been merged.

I'll thus wait for the release branching.

**Comment by gasche** [comment]
I had forgotten about this PR; I would like to work on it more sometime after the release.

**Comment by mshinwell** [comment]
@gasche Would you have time to bring this up-to-date?

**Comment by gasche** [comment]
@mshinwell thanks for your hard triaging work today :-)

**Comment by gasche** [comment]
I rebased my patch against trunk. I should also look at Damien's remark on using in in `typing/Cmi_format.read_cmi` and other places; this is nice but it also increases the invasiveness of the patch a bit.

**Comment by gasche** [comment]
There seems to be a minor double-`close_in` bug in [the current `read_cmi`](https://github.com/ocaml/ocaml/blob/30c7967/typing/cmi_format.ml#L47-L74): it is my impression that the `Error` exception path results in `ic` being closed twice.

**Comment by gasche** [comment]
So I tried to use the new functions in a few places in the compiler -- namely `typing/Cmi_format.read_cmi` and `asmcomp/Compilenv.read_unit_info`. I would be interested in feedback before trying to convert the other use sites -- ideally, I would like an approval to merge in principle.

A difficulty that made this quite more difficult than I expected is that all use-sites have already their own error-reporting logic, and that it is not trivial to integrate the `Magic_number` explicit approach without either changing the error-reporting logic or ending up with code that is roughly as long as what was before (because the concision gain is lost in re-routing exceptions of one kind into another). I had to do several iterations of convenience functions in the `Magic_number` module to end up with something worthwhile.

Changing the error-reporting logic results in better-quality error messages (which was the goal in the first place), but it requires changing the exception types for the affected modules. We have to think about this, because it means that it can break `compiler-libs`-using code. Luckily, the changes made will break programs at compile-time, not runtime.

For each of the two updated sites I thus did two iterations:
- the first one is to just use the functions of `Magic_number`, with no changes to the error reporting logic:
  `read_cmi` is updated in 3f366809f93cb36888b6b53a9a320ccbe53f26c3, and `read_unit_info` in 7eeb8f24870c62da516bddaeffcea7e15a9f17c6
- the second one changes the error-reporting logic to provide consistent information among all `Magic_number` users: `read_cmi` is updated in 2212e4cf7746196ec41f06e736fe3ce862a38f39 (besides regularity we do not gain much as it already supported older/newer recommendations), `read_unit_info` is updated in 47012bb63e91e0e3937d9020e9d30f72c4251b4f.

@damiendoligez, would have a look and give your opinion on whether you think these changes are worthwhile? If we consider that the improved error messages are not worth changing the error type declarations, I can easily roll-back to the "first iteration" approach.

File: utils/misc.ml
Comment (Drup) [inline utils/misc.ml:1026]: Maybe use a list of pairs and avoid duplication ? Matching/printing of magic number is not exactly performance critical.

File: utils/misc.mli
Comment (Drup) [inline utils/misc.mli]: That's very optimistic and hypothetical. ;)

File: utils/misc.ml
Comment (Drup) [inline utils/misc.ml]: If you are raising this, you should register a printer using `Location.register_error_of_exn`

File: typing/cmi_format.ml
Comment (Drup) [inline typing/cmi_format.ml]: You could just keep the structured error (`Magic_number.parse_error`) here.

File: tools/objinfo.ml
Comment (Drup) [inline tools/objinfo.ml]: Why not use that error message for `Magic_number.explanation` ?

**Review by Drup — COMMENTED** [review]
I like the new API quite a lot. As someone who hacked into cmi files at some point. this API would be much easier to tinker with. 

Something that would make it even better (for me) would be to have a function similar to `check_current` but of type `kind list -> info -> (kind, unexpected_error) result`, which would try to check for several magic kind, and return the one we got.

File: typing/cmi_format.ml
Comment (gasche) [reply typing/cmi_format.ml]: I did consider this -- and the `parse_error` structure has grown closed to these module-specific `error` structures over time (in particular I inherited the Corrupted/Not_a_magic_number distinction from them), but I think that this way is better:

- it is more consistent with other exception constructors that tend to often carry "unstructured" rather than "structured" payloads (I'm not for keeping bad practices indefinitely by "coherence in abstraction levels" is good as well)
- this allows me to isolate the `Magic_number`-using code in only one part of the module, the one reading the files. Otherwise I would have `Magic_number` usage spread to both the exception-raisers and exception-catchers.

File: tools/objinfo.ml
Comment (gasche) [reply tools/objinfo.ml]: Good idea.

File: utils/misc.ml
Comment (gasche) [reply utils/misc.ml:1026]: I like the straightforwardness of the present approach, and the duplication is highly local. Besides, I care about `parse_kind`'s performance more than about `raw_kind`, and unfortunately it is easier and safer to define `parse_kind` from `raw_kind` than the other way around.

File: utils/misc.ml
Comment (gasche) [reply utils/misc.ml]: I need to think about this. My intent was that these errors would be handled purely locally by the callers, and not bubble up to a registered exception handler. (If I wanted to let them bubble I should include more information in them to have a chance at an understandable message, such as the (optional) filename they are coming from.) Maybe I should express my intent in the documentation.

I could also remove the exception-raising variants, but in my experience they are important to make the use-sites concise (no chaining of results) and closer to their previous implementation.

**Comment by gasche** [comment]
Thanks @Drup for the review. Could you point me to a codebase with a use-site for your `kind list -> ...` idea?

One thing one has to be careful about is that not all headers are at the same place: the header for `Exec` is to be found at the *end* of the executable file, and the header for `cmxs` files is found at a place determined by `objinfo_helper`. It would thus not be correct to call `expect [Exec; Cmo] ic` for example, and I would rather keep my scope limited to the handling of magic *numbers* rather than all things ocaml-object-files.

File: utils/misc.ml
Comment (gasche) [reply utils/misc.ml:1026]: (Mad feature wish: bidirectional clauses)

File: utils/misc.ml
Comment (Drup) [reply utils/misc.ml]: This is not a high-usage API. Safety > Conciseness for internal compiler APIs, so I would remove the exceptions.

File: utils/misc.ml
Comment (Drup) [reply utils/misc.ml:1026]: Note that you already have a list of the kind, `all_kind`. So you would just extend that.

File: typing/cmi_format.ml
Comment (Drup) [reply typing/cmi_format.ml]: I'm not sure which exception constructors your are thinking of, but my experience is the opposite: trying to keep the most structured error possible until the printer that is registered by `Location.register...`. See `Includemod` for example. I can't even think of an error type with error messages pre-expanded.
So, for once, the common practice is also the good one. :)

Comment (gasche) [reply typing/cmi_format.ml]: Well, for example the `Wrong_version_interface` constructor in the diff you are looking at.

**Comment by Drup** [comment]
@gasche the context is my [eliom compiler](https://github.com/ocsigen/ocaml-eliom). The precise code is a bit messy at the moment, but you can look [here](https://github.com/ocsigen/ocaml-eliom/blob/master/typing/cmi_format.ml#L48).

Basically, the eliom compiler can output cmi files for eliom modules that have more information (encoded in attributes) that would not make sense for the OCaml compiler. In order to avoid confusion, I use a different magic for those "eliom cmi" (even if binary-compatibility is preserved). When reading a cmi. I check the magic and do slightly different action depending if it's a normal cmi or an eliom cmi.

Basically, it allows to output special `cm*` files occasionally, while preserving compatibility with the normal ocaml files.

File: typing/cmi_format.ml
Comment (gasche) [reply typing/cmi_format.ml]: P.S.: I plan to reconsider my position above somewhat, but right now I have zero time to work on this further.

**Comment by mshinwell** [comment]
@gasche Could we try to bring this pull request to a conclusion?

File: typing/cmi_format.ml
Comment (damiendoligez) [reply typing/cmi_format.ml]: What about now?

File: typing/cmi_format.ml
Comment (damiendoligez) [inline typing/cmi_format.ml]: Why not just remove the old code instead of commenting it?

File: utils/misc.ml
Comment (damiendoligez) [inline utils/misc.ml:1093]: Is it really useful to distinguish between `Corrupted` and `Not_a_magic_number`? Note that by your definition and empty file is always "corrupted".

File: utils/misc.ml
Comment (damiendoligez) [inline utils/misc.ml]: Do we really want to run this every time the compiler starts up? If there is a problem with magic numbers, the compiler will certainly fail quite quickly anyway.

File: utils/misc.ml
Comment (damiendoligez) [inline utils/misc.ml]: should be `compiled`

File: utils/misc.ml
Comment (damiendoligez) [inline utils/misc.ml:1058]: Given the way you use these description, I suggest adding the usual file extension in parentheses (where applicable). For example:

```
"bytecode library (.cma)"
```

File: utils/misc.mli
Comment (damiendoligez) [inline utils/misc.mli]: should be `End_of_file`

File: utils/misc.mli
Comment (damiendoligez) [inline utils/misc.mli]: Don't forget to update this.

File: utils/misc.mli
Comment (damiendoligez) [inline utils/misc.mli:597]: `Truncated` should be `Corrupted`.

File: utils/misc.mli
Comment (damiendoligez) [inline utils/misc.mli]: too many `numbers`

**Review by damiendoligez — APPROVED** [review]
Looks good. I've made a few remarks and there is one thing that you wanted to revisit.

**Comment by shindere** [comment]
I just tried to review this PR. In my opinion, it is a quite big bullet to
swallow. So, I am wondering whether it wouldn't be better to split it
in smaller parts. I realise this is far from being trivial and may require
to re-write everything, but I still think it would be interesting both in
terms of distinguishing funcitonalities from each other and for making
reviewing easier.

Regarding how this could be split, I'd suggest to start by just improving
the reportiing of wrong magic numbers, without trying to be more
clever than that.

Perhaps while doing this not-too-difficult and not-too-controversial bit one
can try to do some refactoring, to make sure the way errors are handled is
as uniform as possible.

Then, IMO, becoming clever about the magic numbers and relatiing them to
OCaml versions could be implemented in a
separate PR.

**Comment by Drup** [comment]
I do not really agree with @shindere. Most of the code is contained in a small separated module which abstract away all the relevant functions. These functions are then used in all the relevant places. Also, the commit log is very clean. I don't think it's worth trying to cut the feature in smaller pieces.

**Comment by gasche** [comment]
For the record, this PR was stalling because I haven't taken the time to update it after the feedback from @Drup and @damiendoligez. I'm not sure yet when I'll turn the crank another time, but hopefully soon-ish -- in time for 4.08.

**Comment by shindere** [comment]
Well, to me, the PR contains several features, e.g.(1) checking the
magic numbers and (2) doing some parsing of them. To me, these features
are orthogonal, that's why it would make sense to keep them separate.
For instance, I could very well imagine that, rather than parsing the
magic numbers, one would maintain a map of all the known magic numbers
and the compiler version they are associated with.

**Comment by xavierleroy** [comment]
A lot of development and reviewing work by our finest youth went into this PR, so I'm sad to see it stalling and moving towards oblivion.  On the other hand, that is a lot of code for a small benefit.  Can we decide to bring it to completion or forget about it completely?

**Comment by whitequark** [comment]
In my opinion, this PR is reasonably self-contained, and in my experience, OCaml compiler version mismatches are common in a world with `opam` and quite frustrating to novices. I think merging this PR would be an important improvement to user experience that is "off the beaten path" but also still very common.

**Comment by shindere** [comment]
Xavier Leroy (2019/10/16 09:09 -0700):
> A lot of development and reviewing work by our finest youth went into
> this PR, so I'm sad to see it stalling and moving towards oblivion.
> On the other hand, that is a lot of code for a small benefit.  Can we
> decide to bring it to completion or forget about it completely?

Yeah, I confirm that I still can see a bit of usefulness in this
paragraph.

@xavierleroy Do you really think not providing this bit of information
is any better than providing it, as incomplete as you may find it?


To me there definitely is value in providing a way to automatically test
the bootstrap, especially given the fact that, as far as I know, none of
the current CI solutions that we use on PRs verifies that it has not
been broken. So, currently, this is verified only after a PR gets
merged, which means rather late.

Would there be a rewording or so that would make you like this PR more?

**Comment by xavierleroy** [comment]
@shindere: it looks like your reply belongs to #1819 , not to this PR.

**Comment by shindere** [comment]
Xavier Leroy (2019/10/17 02:41 -0700):
> @shindere: it looks like your reply belongs to #1819 , not to this PR.

My bad. I moved the comment to the right place, thanks.*

**Comment by gasche** [comment]
At the time this PR was written, the magic number of a certain kind of file depended on the year at which the magic number was introduced: most magic numbers started with `1999` (for example `1999{I,O,A}` for `cm{i,o,a}` files), except `cmxs` (`2007D`) and `cmt` (`2012T`).

In October 2017, @damiendoligez changed this to use 1999 everywhere in 8ba70ff194b66c0a50ffb97d41fe9c4bdf9362d6 (discussed in #1402).

It means that certain kinds of files numbers use different magic number prefixes over time. I will adapt the PR to support this.

(Another such change I have to deal with is the fact that flambda uses slightly different kind prefixes for `cmx` and `cmxa` files since 828cb4ebb918bdc4bba167c87e4dd0e3b6f75f38. That commit was merged around the same time the present PR was proposed, but I haven't taken it into account yet.)

**Comment by xavierleroy** [comment]
> the magic number of a certain kind of file depended on the year at which the magic number was introduced: most magic numbers started with 1999

Actually, Caml1999 was one of the code names for what is now OCaml, as a pun on the [ML2000](https://www.researchgate.net/publication/242362854_Principles_and_a_Preliminary_Design_for_ML2000) project. The prefix `Caml1999` was used as early as 1995.  But maybe the time has come to use a more logical scheme for magic numbers, as suggested in #1402 .

File: utils/misc.ml
Comment (gasche) [reply utils/misc.ml:1026]: Note: I kept the existing approach in the rebased PR, as I still believe that straightforwardness and performance matter more than the local duplication.

File: utils/misc.mli
Comment (gasche) [reply utils/misc.mli]: I got rid of this comment.

File: utils/misc.ml
Comment (gasche) [reply utils/misc.ml]: I got rid of this home-grown exception type in favor of using `result` only.

File: typing/cmi_format.ml
Comment (gasche) [reply typing/cmi_format.ml]: In the rebased version, I keep structured errors in exception payloads. (And the structured errors are nicer than before.)

File: typing/cmi_format.ml
Comment (gasche) [reply typing/cmi_format.ml]: done

File: utils/misc.ml
Comment (gasche) [reply utils/misc.ml:1093]: I kept the distinction, but I went back to `Truncated` as the constructor name. The `explain` function now special-cases `Truncated ""` and says that the file is empty (instead of just "truncated").

File: utils/misc.ml
Comment (gasche) [reply utils/misc.ml:1058]: The `explain` function(s) now add this (without the dot, because I think that some names are not file extensions (ast_impl, ast_intf)).

File: utils/misc.mli
Comment (gasche) [reply utils/misc.mli:597]: In the end I went back to `Truncated`. Corrupted is used for files that do have a valid magic number but then fail when parsing the file along the expected format.

**Comment by gasche** [comment]
I just rebased the PR on top of trunk. I also redesigned the API of the Magic_number module, and (following @shindere's comments on feature bloat) restricted the scope of the PR to improving `objinfo.ml` only. I also rebased the changes to Compilenv and Cmi_format to use the new module, to test the new API, but I will keep them for a future PR.

In his review @Drup had made criticism of the API, which relied too much on exceptions. I used exceptions to make caller-site changes less invasive. The new API is written without any exception, purely using the `result` type. I think the new API is sensibly better, but adapting the callsites is now more work. (I buy @Drup's argument that we don't read magic numbers very often in the compiler codebase and can afford the extra verbosity in exchange for more user-friendly failure messages.) In general I went in the direction of making the modified code as good as I could, even if it meant restructuring the code rather than just replacing a couple function calls; the changes to `objinfo.ml` are also a refactoring.

@Drup, would you be willing to give another round of review to the PR?

**Comment by gasche** [comment]
(I managed to convince Travis to indeed check this PR.)

File: tools/objinfo.ml
Comment (Drup) [inline tools/objinfo.ml:393]: I understand your fondness for your newly coined indentation style but it's 1) not particularly standard 2) gives the wrong intuition about how things are nested 3) only reduce indentation depth by 2. Please use the regular style in the compiler. :)

File: utils/misc.mli
Comment (Drup) [inline utils/misc.mli:634]: Why not an optional argument ?

Comment (Drup) [reply utils/misc.mli:634]: Also, the documentation should mention that this behaves as `read_info` wrt the channel.

File: utils/misc.ml
Comment (Drup) [reply utils/misc.ml]: I would tend to agree. This could just be in the testsuite.

File: utils/misc.ml
Comment (Drup) [inline utils/misc.ml:974]: Any good reason not to expose `all_kinds` ?

**Review by Drup — APPROVED** [review]
Looks good! As an added benefit, `objinfo` now understands serialized ASTs, which is nice. All my comments are minor. travis also spotted a few long-lines:
```
./tools/objinfo.ml:334.81: [long-line] line is over 80 columns
./utils/misc.ml:996.81: [long-line] line is over 80 columns
```

File: utils/misc.mli
Comment (gasche) [reply utils/misc.mli:634]: Optional arguments are generally a bad idea for API design, unless we have strong confidence that the default choice is the correct "path of least effort" that will do the right thing for users who don't read the documentation carefully. In this case it's pretty clear that there is no correct default value for this value, and users really need to think about what they should put in this argument.

In fact an intermediary state of this PR had two separate functions, one for reading a "current" number at any possible kind, and one for reading a current number at a specific expected kind. I think that the present approach is better (at least on the implementation side it reduces duplication), but it's important that the choice be explicit.

(Good point on the documentation.)

File: utils/misc.ml
Comment (gasche) [reply utils/misc.ml:974]: For now I don't see a reason to expose it. If people come up with a good reason later, I'm happy for it to be exposed as a further change.

File: utils/misc.mli
Comment (gasche) [reply utils/misc.mli:634]: The unspoken assumption that any labelled argument of option type could be made optional is bad and we should get rid of it!

File: utils/misc.ml
Comment (gasche) [reply utils/misc.ml]: I moved this into the testsuite. This required exposing previously-private parts of the module that were used by the test. I'm not completely satisfied with this but I think that it's the best compromise on this point.

File: utils/misc.mli
Comment (gasche) [reply utils/misc.mli:634]: I updated the documentation to mention `read_info`.

**Comment by gasche** [comment]
I rebased the PR to address @Drup's last batch of comments.

**Comment by gasche** [comment]
I went ahead and merged in trunk. Thanks again @damiendoligez and @Drup for the reviews.

This story arc is not yet complete; only `ocamlobjinfo` uses the new machinery as of now, all other parts of the compiler that read magic numbers could be improved by adopting it as well. I (and maybe others as well?) will send extra PRs over time.

**Comment by shindere** [comment]
The objinfo test is now broken on Wygwin32.

File question.cmxs
Error: size of section .data unknown
Unable to read info on file question.cmxs

**Comment by shindere** [comment]
And by the way, this also affects the 4.10 branch.

**Comment by gasche** [comment]
Do you think that it is related to this PR? This PR was merged two days ago, in trunk only (I think? no reason to cherry-pick to 4.10), so I don't think it could explain a broken test on 4.10.

**Comment by shindere** [comment]
Gabriel Scherer (2019/12/05 01:52 -0800):
> Do you think that it is related to this PR? This PR was merged two
> days ago, in trunk only (I think? no reason to cherry-pick to 4.10),
> so I don't think it could explain a broken test on 4.10.

Well that may be a wrong guess of mine, sorry if that's the case.

Given the times it takes to build OCaml on Cygwin the bisection is not a
very appealing solution so I guess somebody will have to investigate.
I'll see whether I can do it or not.

**Comment by gasche** [comment]
(I think that today, being a general strike day in France, is an excellent day to *not* investigate an obscure Cygin32 failure (don't count on me!), but I'm going to humor you and ask the first question that comes to mind.) Do we know when the failures started to happen?

**Comment by shindere** [comment]
Gabriel Scherer (2019/12/05 02:36 -0800):
> (I think that today, being a general strike day in France, is an
> excellent day to *not* investigate an obscure Cygin32 failure (don't
> count on me!), but I'm going to humor you and ask the first question
> that comes to mind.) Do we know when the failures started to happen?

No, we don't.

And actually I am now investigating another hypothesis. It may very well
be that this test was actually never executed under Cygwin32 because its
BFD library was just not detected. Due to the recent fixes in the BFD
support, hte library is now correctly detected and the test is run...
and fails.

**Comment by shindere** [comment]
I confirm the hypothesis mentionned above.

objinfo_helper can not find the caml_header_plugin symbol for several
reasons. I have a dirty but working fix. Still discussing off-line to come
up with a cleaner one, will open a new PR ASAP. Sorry for the false positive
here.

**Pull request merged**
